File: day-3/python/parser.py
"""Module providing utilities for parsing input data."""
import typing

from wire import Point, Wire


def parse(instructions: typing.List[str]) -> Wire:
    """Calculate the points upon a wire and create a class from them."""
    points = [Point(0, 0)]

    for instruction in instructions:
        dir = instruction[:1]
        dist = int(instruction[1:])

        for _ in range(dist):
            if dir == "R":
                next_point = points[-1].right()
            elif dir == "L":
                next_point = points[-1].left()
            elif dir == "U":
                next_point = points[-1].up()
            else:
                next_point = points[-1].down()

            points.append(next_point)

    return Wire(points)

File: day-3/rust/src/parser.rs
use super::wire::{Wire, Point};

pub fn parse(instructions: &str) -> Wire {
    let inst = instructions.split(",").map(|x| x.to_string()).collect::<Vec<String>>();


    let mut next_point = Point::new(0, 0);
    let mut points: Vec<Point> = vec![];

    for instruction in inst {
        let dir = &instruction[..1];
        let dist = &instruction[1..].parse::<i64>()
          .expect(&format!("Could not parse integer of {}", instruction));

        for _ in 0..*dist {
            match dir {
                "U" => next_point.up(1),
                "D" => next_point.down(1),
                "L" => next_point.left(1),
                "R" => next_point.right(1),
                _ => panic!("Unexpected direction: {}", dir)
            };
            points.push(next_point.clone());
        }
    };

    Wire::new(points)
}
